src/core/worker.lua

```diff
@@ -0,0 +1,132 @@
+
+module(..., package.seeall)
+
+-- API:
+-- start(name, luacode)
+-- stop(name)
+-- status() -> table of { name = <info> }
+
+local lib = require("core.lib")
+local shm = require("core.shm")
+local S = require("syscall")
+
+--------------------------------------------------------------
+-- Master (parent) process code
+--------------------------------------------------------------
+
+local children = {}
+
+local function child (name)
+   return children[name] or error("no such child: " .. name)
+end
+
+-- Start a named worker to execute the given Lua code (a string).
+function start (name, luacode)
+   local pid = S.fork()
+   if pid == 0 then
+      -- First we perform some initialization functions and then we
+      -- restart the process with execv().
+      
+      -- Terminate automatically when the parent dies.
+      --
+      -- XXX This prctl setting needs to survive execve(). The Linux
+      -- execve(2) page seems to say that it will provided that the
+      -- binary being executed is not setuid or setgid. This may or
+      -- may not be adequate.
+      S.prctl("set_pdeathsig", "hup")
+      -- Symlink the shm "group" folder to be shared via the parent process.
+      shm.alias("group", "/"..S.getppid().."/group")
+      -- Save the code we want to run in the environment.
+      S.setenv("SNABB_PROGRAM_LUACODE", luacode, true)
+      -- Restart the process with execve().
+      -- /proc/$$/exe is a link to the same Snabb executable that we are running
+      local env = {}
+      for key, value in pairs(S.environ()) do
+         table.insert(env, key.."="..value)
+      end
+      S.execve(("/proc/%d/exe"):format(S.getpid()), {}, env)
+   else
+      -- Parent process
+      children[name] = { pid = pid }
+   end
+end
+
+-- Terminate a child process
+function stop (name)
+   S.kill(child(name).pid, 'kill')
+end
+
+-- Return information about all worker processes in a table.
+function status ()
+   local status = {}
+   for name, info in pairs(children) do
+      local infop = S.waitid("pid", info.pid, "nohang, exited")
+      status[name] = {
+         pid = info.pid,
+         alive = infop and infop.code == 0 or false
+      }
+   end
+   return status
+end
+
+--------------------------------------------------------------
+-- Worker (child) process code
+--------------------------------------------------------------
+
+-- Initialize the worker by attaching to relevant shared memory
+-- objects and entering the main engine loop.
+function init ()
+   local name = assert(lib.getenv("SNABB_WORKER_NAME"))
+   local parent = assert(lib.getenv("SNABB_WORKER_PARENT"))
+   print(("Starting worker %s for parent %d"):format(name, parent))
+
+   -- Create "group" alias to the shared group folder in the parent process
+   shm.alias("group", "/"..parent.."/group")
+
+   -- Run the engine with continuous configuration updates
+   local current_config
+   local child_path = "group/config/..name"
```

Comment:
ahem! That whole `init()` function is actually dead code. Removed in 876addc.

That's from an older version where `core.worker` imposed specific behavior on the worker process i.e. running an app network provided by the parent. These days the worker simply links itself into the process tree and runs the provided Lua code.